config/config.go

```diff
@@ -1015,6 +1047,70 @@ func configFileHasDependencyBlock(configPath string, terragruntOptions *options.
 	return false, nil
 }
 
+const (
+	MetadataTerraform                   = "terraform"
+	MetadataTerraformBinary             = "terraform_binary"
+	MetadataTerraformVersionConstraint  = "terraform_version_constraint"
+	MetadataTerragruntVersionConstraint = "terragrunt_version_constraint"
+	MetadataRemoteState                 = "remote_state"
+	MetadataDependencies                = "dependencies"
+	MetadataDependency                  = "dependency"
+	MetadataDownloadDir                 = "download_dir"
+	MetadataPreventDestroy              = "prevent_destroy"
+	MetadataSkip                        = "skip"
+	MetadataIamRole                     = "iam_role"
+	MetadataIamAssumeRoleDuration       = "iam_assume_role_duration"
+	MetadataIamAssumeRoleSessionName    = "iam_assume_role_session_name"
+	MetadataInputs                      = "inputs"
+	MetadataLocals                      = "locals"
+	MetadataGenerateConfigs             = "generate"
+	MetadataRetryableErrors             = "retryable_errors"
+	MetadataRetryMaxAttempts            = "retry_max_attempts"
+	MetadataRetrySleepIntervalSec       = "retry_sleep_interval_sec"
+)
+
+// AddFieldMetadata persist "generic" metadata
+func (conf *TerragruntConfig) MetadataWithType(fieldType, fieldName string, m map[string]interface{}) {
```

Comment:
NIT: The name `MetadataWithType` makes me think of _type constraints_: e.g., string, int, bool, etc. I don't think that's what's happening here... Is this more of there being a _fieldName_ (the name of a field, which is a constant) and _fieldId_ (a unique ID)? And so the method signature is more like `SetMetadataWithId(fieldName, fieldId string, ...)`.